Rework for nested `AnimationNodeStateMachine`

TokageItLab opened the pull request:
~~Prerequisite #75111.~~ Now it has been included in this.

- Fixed #71562

This PR replaces/removes the previous broken nested StateMachine behavior. Previously, StateMachine stored pointers to AnimationNode within child StateMachines. However, storing pointers to other Resources within a Resource was unsafe because it could lead to duplicated processing within the animation blending process. It was causing problems similar to #22887.

This PR does not store pointers directly, but generates paths to PlaybackObject so that the child (or parent) state machine pointers are accessed only at the moment of the blending process.

- Fixed #58822
- Fixed #73643

Also, make the PlaybackObject readonly and it is not stored in tres. This prevents the PlaybackObject from having the same reference, since the PlaybackObject is created with a unique value each time.

- Fixed #45635

## Make the StateMachine to calculate the remain time at the transition to End state

Until now, the remain time was always treated as 1 second until the End state was reached. This fix will allow the fade-out to be handled correctly when using  the StateMachine from the outside such as BlendTree.

- Fixed #73648
- Fixed #74752

---

## Implement three types of StateMachine for different use cases

### RootStateMachine

This is the default state machine. The most standard StateMachine, applicable to most use cases.

![image](https://user-images.githubusercontent.com/61938263/230419411-1162bb30-6a15-4e97-96fd-fb5ae3e175aa.png)

When a seek to the beginning is commanded, a transition is made to StartState.

https://user-images.githubusercontent.com/61938263/230428527-1adfb466-6c13-4ae5-b99a-e95926e00e42.mp4

- Fixed #57223
- Fixed #68986
- Fixed #69063

And remain time is calculated only when transitioning to End state.

### NestedStateMachine

It is useful to list several animations, such as StateMnachine with NodeTransition. You can use this like AnyState. Seeking to the beginning is treated as seeking to the beginning of the animation in the current state.

![image](https://user-images.githubusercontent.com/61938263/230419516-d0798177-e248-4244-9768-06b9a9cbbf21.png)

The remain time is calculated not only at the transition to End state, but also when the animation is stopped due to a state with no transition.

### GroupedStateMachine

This is useful for organizing the StateMachine so that it does not make it too messy.

With some restrictions, GroupedStateMachine allows direct travel between StateMachines.

https://user-images.githubusercontent.com/61938263/230428319-227de9b8-778a-43b9-a54b-0bb913eee97b.mp4

- Fixed #62576.

#### Restrictions

Users cannot directly request play/travel to a StateMachinePlayback of GroupedStateMachine, since GroupedStateMachine must be processed by the parent StateMachinePlayback.

This means that a GroupedStateMachine must have a Root/Nested StateMachine as a parent or ancestor.

For example, you can play/travel in the parent Root/Nested StateMachine as follows:

```gdscript
root_playback.travel("Combo/Combo1")
```

The Start/End transition of a GroupedStateMachine does not allow parameters to be set. This is because the transitions are referenced as shown in the following figure:

![image](https://user-images.githubusercontent.com/61938263/230419723-9cc29317-a20d-4cd1-99d6-61848bbace33.png)

Then, the number of transitions to/from the GroupedStateMachine and the number of internal Start/End transitions must be matched.

It is also recommended that no more than one transition be made to a GroupedStateMachine. This is because there is currently only one input port to the StateMachine, so if there are two or more transitions, it is not possible to specify the reference for each.

However, this problem can be solved in the future by improving the StateMachine to have more than two input ports, or by increasing the number of StartNodes, etc.

Incidentally, to ensure that the reference to transition is correct, direct transitions to the Start or End State of a GroupedStateMachine is not allowed.

```gdscript
root_playback.travel("Combo/Start") # It's not alllowed.
```

However, for convenience, if a travel is specified to a GroupedStateMachine, it is implicitly replaced by a travel to the next State after Start if a Start transition exists.

```gdscript
root_playback.travel("Combo")
root_playback.travel("Combo/Combo1") # Same result with above.
```

And, of course, if the priority depend on the ASter algorithm, the equality of travel path is not guaranteed before and after they are grouped.

## TODO in the future

- Update document
- Optimize to cache parent-child relationships for path retrieval as like #75627
- Increase the number of input ports or StartNodes for grouped state machines
- Allow grouping using the GUI but the priority of this is low because the equality of transitions before and after can't be guaranteed without doing above (increasing port) first

[state_machine_improvement_demo.zip](https://github.com/godotengine/godot/files/13993286/state_machine_improvement_demo.zip)
[state_machine_improvement_demo_2.zip](https://github.com/godotengine/godot/files/13993496/state_machine_improvement_demo_2.zip)

TokageItLab committed:
Rework StateMachine and nested StateMachine process

Breaking compatibility: If a StateMachineTransition is connected to a nested StateMachine prior to this, it is removed. Also, there was a feature to connect another StateMachine as the End of a StateMachine, which is also removed to avoid reference confusion. It was like a GoTo Statement, also further passing the same reference to the blending process, causing the blending calculation to break or causing some StateMachines to not work.

fire commented:
Lyuma is testing nested nested state machines. We are discussing off github.

alfredbaudisch commented:
Is this one also related https://github.com/godotengine/godot/issues/70338?

TokageItLab commented:
@alfredbaudisch No, https://github.com/godotengine/godot/issues/70338 is related https://github.com/godotengine/godot-proposals/issues/5972. This PR is proposal of https://github.com/godotengine/godot-proposals/issues/6130

SaracenOne reviewed (approved):
Sorry this took so long, but yes, having looked through the code and experimented with it, it looks good to me! 👍

YuriSizov commented:
Commits need to be squashed. And could you please do a write-up of how the compatibility is broken in this PR and how users should handle it in their projects?

TokageItLab commented:
Squashed.

---

Breaking compatibility:

- If a StateMachineTransition is connected to a nested StateMachine prior to this, it is removed.
- Also, there was a feature to connect another StateMachine as the End of a StateMachine, which is also removed to avoid reference confusion.

Since it was like a GoTo Statement, also further passing the same reference to the blending process, causing the blending calculation to break or causing some StateMachines to not work.

akien-mga commented:
Thanks!
